src/util/mod.rs

```diff
@@ -347,16 +430,75 @@ pub fn file_icon_suffix<P: AsRef<Path>>(f: P, is_file: bool) -> &'static str {
 ///
 /// The specified path must point to a file.
 pub fn get_raw_fs_metadata<P: AsRef<Path>>(f: P) -> RawFileData {
-    let f = f.as_ref();
+    get_raw_fs_metadata_impl(f.as_ref())
+}
+
+fn get_raw_fs_metadata_impl(f: &Path) -> RawFileData {
+    let meta = f.metadata().expect("Failed to get requested file metadata");
     RawFileData {
         mime_type: guess_mime_type_opt(f).unwrap_or_else(|| if file_binary(f) {
             "application/octet-stream".parse().unwrap()
         } else {
             "text/plain".parse().unwrap()
         }),
         name: f.file_name().unwrap().to_str().expect("Failed to get requested file name").to_string(),
-        last_modified: file_time_modified(f),
-        size: f.metadata().expect("Failed to get requested file metadata").len(),
+        last_modified: file_time_modified(&meta),
+        size: file_length(&meta, &f),
         is_file: true,
     }
 }
+
+/// Recursively copy a directory
+///
+/// Stolen from https://github.com/mdunsmuir/copy_dir/blob/0.1.2/src/lib.rs
+pub fn copy_dir(from: &Path, to: &Path) -> IoResult<Vec<(IoError, String)>> {
+    macro_rules! push_error {
+        ($vec:ident, $path:ident, $expr:expr) => {
+            match $expr {
+                Ok(_) => (),
+                Err(e) => $vec.push((e, $path.to_string_lossy().into_owned())),
+            }
+        };
+    }
+
+    let mut errors = Vec::new();
+
+    fs::create_dir(&to)?;
+
+    // The approach taken by this code (i.e. walkdir) will not gracefully
+    // handle copying a directory into itself, so we're going to simply
+    // disallow it by checking the paths. This is a thornier problem than I
+    // wish it was, and I'd like to find a better solution, but for now I
+    // would prefer to return an error rather than having the copy blow up
+    // in users' faces. Ultimately I think a solution to this will involve
+    // not using walkdir at all, and might come along with better handling
+    // of hard links.
+    if from.canonicalize().and_then(|fc| to.canonicalize().map(|tc| (fc, tc))).map(|(fc, tc)| tc.starts_with(fc))? {
+        fs::remove_dir(&to)?;
+
+        return Err(IoError::new(IoErrorKind::Other, "cannot copy to a path prefixed by the source path"));
+    }
+
+    for entry in WalkDir::new(&from).min_depth(1).into_iter().flatten() {
+        let source_metadata = match entry.metadata() {
+            Ok(md) => md,
+            Err(err) => {
+                errors.push((err.into(), entry.path().to_string_lossy().into_owned()));
+                continue;
+            }
+        };
+
+        let relative_path = entry.path().strip_prefix(&from).expect("strip_prefix failed; this is a probably a bug in copy_dir");
+
+        let target_path = to.join(relative_path);
+
+        if !is_actually_file(&source_metadata.file_type()) {
```

Comment:
We're going to push an error either way? Maybe I just need to work through reading this code a bit more

Reply:
Not sure what this comment means tbh

Reply:
you do `if !test() { foo() } else { bar() }` whilst `if test() { bar() } else { foo() }` makes more sense to me, why negate the Boolean?

Reply:
That's how it was structured originally — https://github.com/mdunsmuir/copy_dir/blob/0.1.2/src/lib.rs#L143.